air-script-core/src/variable.rs

```diff
@@ -25,4 +25,51 @@ pub enum VariableType {
     Scalar(Expression),
     Vector(Vec<Expression>),
     Matrix(Vec<Vec<Expression>>),
+    ListComprehension(ListComprehension),
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct ListComprehension {
+    expression: Expression,
+    members: Members,
+    iterables: Iterables,
+}
+
+impl ListComprehension {
+    /// Creates a new list comprehension.
+    pub fn new(expression: Expression, members: Members, iterables: Iterables) -> Self {
+        Self {
+            expression,
+            members,
+            iterables,
+        }
+    }
+
+    /// Returns the expression that is evaluated for each member of the list.
+    pub fn expression(&self) -> &Expression {
+        &self.expression
+    }
+
+    /// Returns the members of the list.
+    pub fn members(&self) -> &Members {
+        &self.members
+    }
+
+    /// Returns the iterables used to generate the list.
+    pub fn iterables(&self) -> &Iterables {
+        &self.iterables
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum Members {
+    Single(Identifier),
+    Double(Identifier, Identifier),
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum Iterables {
+    Single(Identifier),
+    Double(Identifier, Identifier),
+    Range(Range, Identifier),
 }
```

Comment:
Is this a temporary simplification? Ideally, we should not put a limit on the number of iterable lists and be able to do something like:
```
for (x, y, z) in (a, b[1..4], 0..3)
```
And I'm not sure this model will support this.

Reply:
Makes sense. It wasn't meant to be a temporary simplification as I thought this was the extent of the kinds of expressions that we wanted to allow. However, if we want to allow more complicated expressions here, you are right that we should implement a more generic structure and it probably makes sense to do it now instead of coming back to it later so I'll modify this PR.

Reply:
I think at the parsing level the extra complexity would still be pretty minimal. We could have an `Iterable` enum which would describe different variants of what an iterable could be (i.e., identifier, range, identifier + range). And then `ListComprehension` struct could look something like:
```Rust
pub struct ListComprehension {
    expression: Expression,
    context: Vec<(Identifier, Iterable)>,
}
```
Where tuples stored in `context` contain bindings of names to their respective interables. For example, in `for (x, y, z) in (a, b[1..4], 0..3)`, the context would be:
```
(x, Iterable(a)), (y, Iterable(a, 1..4)), (z, Iterable(0..3))
```